byway/src/main/java/com/google/sps/servlets/StopsServlet.java

```diff
@@ -26,23 +42,41 @@
 /** Servlet that handles data for stops */
 @WebServlet("/api/stop")
 public final class StopsServlet extends HttpServlet {
-  @GuardedBy("this")
-  private ArrayList<String> stops= new ArrayList<String>();
   private final Gson gson = new Gson(); 
-  // Implement datastore in routepage branch
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final Type ARRAYLIST_STRING = new TypeToken<ArrayList<String>>() {}.getType();
 
+  /* Passes saved destinations stops (if any) to be shown in the schedule panel */
   @Override
-  public synchronized void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Gson gson = new Gson();
-    String json = gson.toJson(stops);
-    
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // TODO: Replace ID filtering with .get(Key)
+    // Duplicate code was just kept for this reason
+    Filter propertyFilter = new FilterPredicate("id", FilterOperator.EQUAL, 1234);
+    Query query = new Query(Trip.KIND).setFilter(propertyFilter);
+    PreparedQuery results = datastore.prepare(query);
+    List<String> stops= new ArrayList<>();
+
+    // Should only be one entity with this ID
+    for(Entity entity: results.asIterable()){
+       stops = (ArrayList<String>) entity.getProperty("destinations");
+    }
     response.setContentType("application/json;");
-    response.getWriter().println(json);
+    response.getWriter().println(gson.toJson(stops));
   }
 
+  /* Modifies the destinations array of Trip entity in datastore */
   @Override
-  public synchronized void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String stop = request.getParameter("text");
-    stops.add(stop);
-  }
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
+    String stopsAsJSON = request.getParameter("stops");
+
+    ArrayList<String> stops = gson.fromJson(stopsAsJSON, ARRAYLIST_STRING);
+    Filter propertyFilter = new FilterPredicate("id", FilterOperator.EQUAL, 1234);
+    Query query = new Query(Trip.KIND).setFilter(propertyFilter);
+    PreparedQuery results = datastore.prepare(query);
+
+    for(Entity entity: results.asIterable()){
+      entity.setProperty("destinations", stops);
+      datastore.put(entity);
+    }
```

Comment:
The comment above is also for the purpose of not iterating through all results, since it should only find one unique trip. I have an implementation using this in my `store-interests` branch using the hard-coded id value for the trip if you want to take a look: https://github.com/googleinterns/step200-2020/pull/18/commits/9a96386d9b991e42e3eb325d83c44317756ab56c

Reply:
Yeah, my only issue with that is the hardcoded ID we made is not used, we have to still rely on the randomly generated ID generated at creation of the entity. But I don't think this is a huge deal since we are planning to use the Trip keys directly, so I will change for now.